crates/pathfinder/src/merkle_tree/node.rs

```diff
@@ -0,0 +1,574 @@
+    /// The hash of this node. Is [None] if the node
+    /// has not yet been committed.
+    pub hash: Option<StarkHash>,
+    /// The starting height of this node in the tree.
+    pub height: usize,
+    /// The path this edge takes.
+    pub path: BitVec<Msb0, u8>,
+    /// The child of this node.
+    pub child: Rc<RefCell<Node>>,
+}
+
+/// Describes the direction a child of a [BinaryNode] may have.
+///
+/// Binary nodes have two children, one left and one right.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Direction {
+    Left,
+    Right,
+}
+
+impl Direction {
+    /// Inverts the [Direction].
+    ///
+    /// [Left] becomes [Right], and [Right] becomes [Left].
+    ///
+    /// [Left]: Direction::Left
+    /// [Right]: Direction::Right
+    pub fn invert(self) -> Direction {
+        match self {
+            Direction::Left => Direction::Right,
+            Direction::Right => Direction::Left,
+        }
+    }
+}
+
+impl From<bool> for Direction {
+    fn from(tf: bool) -> Self {
+        match tf {
+            true => Direction::Right,
+            false => Direction::Left,
+        }
+    }
+}
+
+impl From<Direction> for bool {
+    fn from(direction: Direction) -> Self {
+        match direction {
+            Direction::Left => false,
+            Direction::Right => true,
+        }
+    }
+}
+
+impl BinaryNode {
+    /// Maps the key's bit at the binary node's height to a [Direction].
+    ///
+    /// This can be used to check which direction the key descibes in the context
+    /// of this binary node i.e. which direction the child along the key's path would
+    /// take.
+    pub fn direction(&self, key: StarkHash) -> Direction {
+        key.view_bits()[self.height].into()
+    }
+
+    /// Returns the [Left] or [Right] child.
+    ///
+    /// [Left]: Direction::Left
+    /// [Right]: Direction::Right
+    pub fn get_child(&self, direction: Direction) -> Rc<RefCell<Node>> {
+        match direction {
+            Direction::Left => self.left.clone(),
+            Direction::Right => self.right.clone(),
+        }
+    }
+
+    /// If possible, calculates and sets its own hash value.
+    ///
+    /// Does nothing if the hash is already [Some].
+    ///
+    /// If either childs hash is [None], then the hash cannot
+    /// be calculated and it will remain [None].
+    pub fn calculate_hash(&mut self) {
+        if self.hash.is_some() {
+            return;
+        }
+
+        let left = match self.left.borrow().hash() {
+            Some(hash) => hash,
+            None => return,
+        };
+
+        let right = match self.right.borrow().hash() {
+            Some(hash) => hash,
+            None => return,
+        };
+
+        self.hash = Some(pedersen_hash(left, right));
+    }
+}
+
+impl Node {
+    /// Convenience function which sets the inner node's hash to [None], if
+    /// applicable.
+    ///
+    /// Used to indicate that this node has been mutated.
+    pub fn mark_dirty(&mut self) {
```

Comment:
I was actually thinking of having it be an enum to be more explicit.
```rust
enum DirtyOrHash {
    Hash(StarkHash),
    Dirty,
}
```

Reply:
The advantage of the `is_dirty` solution is explicitness, but also allowing mutations to the tree with minimal change (for each dirty node: a bool set to false, rather than zero/reset the entire 32-byte hash). And then when/**_IF_** you commit, you would call a `compute_hash` function that traverse the tree and recompute the necessary hashes.

But for a modern 64-bit SSE/AVX-based machine this is little to no overhead anyway.

Reply:
Oh yeah I'm not concerned with performance at all.